File: Java/SwapPairsInt.java
import static org.junit.Assert.assertEquals;

import java.util.Arrays;

import org.junit.Test;

public class SwapPairsInt {

    public static int reversePairsIntWithClass(int num) {
        int result = 0;
        int i = 0;
        Digits digits = new Digits(num);

        while (digits.getNumber() > 10) {
            result += digits.pop() * Math.pow(10, i+1);
            result += digits.pop() * Math.pow(10, i);
            i += 2;
        }

        if (num > 0) {
            result += digits.pop() * Math.pow(10, i);
        }

        return result;
    }

    public static int reversePairsInt(int num) {
        int result = 0;
        int i = 0;
        while (num > 10) {
            int right_digit = num % 10;
            num = num / 10;

            int left_digit = num % 10;
            num = num / 10;

            result += left_digit * Math.pow(10, i++);
            result += right_digit * Math.pow(10, i++);
        }

        if (num > 0) {
            int last_digit = num % 10;
            result += last_digit * Math.pow(10, i);
        }

        return result;
    }

    public static void main(String args[]) {
        Arrays
            .asList(args)
            .stream()
            .mapToInt(Integer::parseInt)
            .map(SwapPairsInt::reversePairsInt)
            .forEach(System.out::println);
    }

    @Test
    public void testEvenNumberOfLettersWithClass() {
        assertEquals(reversePairsIntWithClass(1234), 2143);
    }

    @Test
    public void testOddNumberOfLettersWithClass() {
        assertEquals(reversePairsIntWithClass(12345), 13254);
    }

    @Test
    public void testEvenNumberOfLetters() {
        assertEquals(reversePairsInt(1234), 2143);
    }

    @Test
    public void testOddNumberOfLetters() {
        assertEquals(reversePairsInt(12345), 13254);
    }
}

class Digits {
    private int num;

    public Digits(int num) {
        this.num = num;
    }

    public int getNumber() {
        return num;
    }

    public int peak() {
        return this.num % 10;
    }

    public int pop() {
        int digit = this.num % 10;
        this.num = this.num / 10;
        return digit;
    }

}

File: Python/swap_pairs_int.py
import math

def pop(num: int) -> tuple[int, int]:
    digit = num % 10
    new_num = math.floor(num / 10)

    return new_num, digit


def get_length(num: int) -> int:
    if num == 0:
        return 0
    return math.floor(math.log10(num)) + 1


def push_left(num, digit) -> int:
    num_digits = get_length(num)

    return num + digit * 10 ** num_digits


def swap_pairs_int(num: int) -> int:
    result = 0
    while get_length(num) > 1:
        num, right_digit = pop(num)
        num, left_digit = pop(num)

        result = push_left(result, left_digit)
        result = push_left(result, right_digit)

    if get_length(num) == 1:
        _, digit = pop(num)
        result = push_left(result, digit)

    return result



def test_happy_path():
    assert swap_pairs_int(1234) == 2143

def test_odd_length():
    assert swap_pairs_int(123) == 132

if __name__ == "__main__":
    test_happy_path()
